# Customer Segmentation in a Lisbon Hotel Chain

The objective is to explore the historical customer information of a 4Star Hotel in Lisbon in order to segment customers and discover the distinguishing features of each group.  
  
This should allow market to have a better understanding of customers groups in order to better engage with the customer. These informations may impact several areas of interaction with the customer, eg:
Marketing: channels, timings, reinforcement points, selling points,...
Sales: Pricing , customer value,...
Reception: Types of interaction,...



## DataSet Description

Talk about the variables


## Setup and Import

In [4]:
import os
import csv
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import category_encoders as ce
import collections
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA

Data Loading and Initial Analysis

In [5]:
ds = pd.read_csv(r'data\dataset.csv', sep=";")
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111733 entries, 0 to 111732
Data columns (total 29 columns):
ID                      111733 non-null int64
Nationality             111733 non-null object
Age                     107561 non-null float64
DaysSinceCreation       111733 non-null int64
NameHash                111733 non-null object
DocIDHash               110732 non-null object
AverageLeadTime         111733 non-null int64
LodgingRevenue          111733 non-null float64
OtherRevenue            111733 non-null float64
BookingsCanceled        111733 non-null int64
BookingsNoShowed        111733 non-null int64
BookingsCheckedIn       111733 non-null int64
PersonsNights           111733 non-null int64
RoomNights              111733 non-null int64
DistributionChannel     111733 non-null object
MarketSegment           111733 non-null object
SRHighFloor             111733 non-null int64
SRLowFloor              111733 non-null int64
SRAccessibleRoom        111733 non-null int64
SRMe

In [6]:
# Display top 10 rows transposed to show all columns
ds.head(10).transpose()

0  \
ID                                                                    1   
Nationality                                                         PRT   
Age                                                                  52   
DaysSinceCreation                                                   440   
NameHash              0x2C371FD6CE12936774A139FD7430C624F1C4D5109CE6...   
DocIDHash             0x434FD3D59469C73AFEA087017FAF8CA2296493AEABDE...   
AverageLeadTime                                                      59   
LodgingRevenue                                                      292   
OtherRevenue                                                       82.3   
BookingsCanceled                                                      1   
BookingsNoShowed                                                      0   
BookingsCheckedIn                                                     2   
PersonsNights                                                         6   
RoomNights                                                            4   
DistributionChannel                                           Corporate   
MarketSegment                                                 Corporate   
SRHighFloor                                                           0   
SRLowFloor                                                            0   
SRAccessibleRoom                                                      0   
SRMediumFloor                                                         0   
SRBathtub                                                             0   
SRShower                                                              0   
SRCrib                                                                0   
SRKingSizeBed                                                         0   
SRTwinBed                                                             0   
SRNearElevator                                                        0   
SRAwayFromElevator                                                    0   
SRNoAlcoholInMiniBar                                                  0   
SRQuietRoom                                                           0   

                                                                      1  \
ID                                                                    2   
Nationality                                                         PRT   
Age                                                                 NaN   
DaysSinceCreation                                                  1385   
NameHash              0x198CDB98BF37B6E23F9548C56A88B00912D65A9AA0D6...   
DocIDHash             0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
AverageLeadTime                                                      61   
LodgingRevenue                                                      280   
OtherRevenue                                                         53   
BookingsCanceled                                                      0   
BookingsNoShowed                                                      0   
BookingsCheckedIn                                                     1   
PersonsNights                                                        10   
RoomNights                                                            5   
DistributionChannel                               Travel Agent/Operator   
MarketSegment                                     Travel Agent/Operator   
SRHighFloor                                                           0   
SRLowFloor                                                            0   
SRAccessibleRoom                                                      0   
SRMediumFloor                                                         0   
SRBathtub                                                             0   
SRShower                                                              0   
SRCrib                                                                0   
SRKingSizeBed      

In [14]:
# Summary statistics for numerical variables
summary=ds.describe(exclude=[np.object]).transpose()   # exclude only objects, they will be treated next
                                                       # transpose the summary for easier reading
summary.head(len(summary))

count          mean           std   min      25%  \
ID                    111733.0  55867.000000  32254.683151   1.0  27934.0   
Age                   107561.0     45.639191     17.244952 -10.0     33.0   
DaysSinceCreation     111733.0    595.026599    374.657382  36.0    288.0   
AverageLeadTime       111733.0     60.833147     85.115320  -1.0      0.0   
LodgingRevenue        111733.0    283.851283    379.131556   0.0      0.0   
OtherRevenue          111733.0     64.682802    123.580715   0.0      0.0   
BookingsCanceled      111733.0      0.002282      0.080631   0.0      0.0   
BookingsNoShowed      111733.0      0.000600      0.028217   0.0      0.0   
BookingsCheckedIn     111733.0      0.737607      0.730889   0.0      0.0   
PersonsNights         111733.0      4.328318      4.630739   0.0      0.0   
RoomNights            111733.0      2.203825      2.301637   0.0      0.0   
SRHighFloor           111733.0      0.042512      0.201755   0.0      0.0   
SRLowFloor            111733.0      0.001307      0.036125   0.0      0.0   
SRAccessibleRoom      111733.0      0.000224      0.014957   0.0      0.0   
SRMediumFloor         111733.0      0.000770      0.027733   0.0      0.0   
SRBathtub             111733.0      0.003132      0.055881   0.0      0.0   
SRShower              111733.0      0.001629      0.040327   0.0      0.0   
SRCrib                111733.0      0.016181      0.126173   0.0      0.0   
SRKingSizeBed         111733.0      0.363268      0.480943   0.0      0.0   
SRTwinBed             111733.0      0.156811      0.363624   0.0      0.0   
SRNearElevator        111733.0      0.000331      0.018195   0.0      0.0   
SRAwayFromElevator    111733.0      0.003598      0.059874   0.0      0.0   
SRNoAlcoholInMiniBar  111733.0      0.000197      0.014031   0.0      0.0   
SRQuietRoom           111733.0      0.087718      0.282886   0.0      0.0   

                          50%      75%        max  
ID                    55867.0  83800.0  111733.00  
Age                      47.0     58.0     123.00  
DaysSinceCreation       522.0    889.0    1385.00  
AverageLeadTime          21.0     95.0     588.00  
LodgingRevenue          208.0    393.3   21781.00  
OtherRevenue             31.0     84.0    8859.25  
BookingsCanceled          0.0      0.0      15.00  
BookingsNoShowed          0.0      0.0       3.00  
BookingsCheckedIn         1.0      1.0      76.00  
PersonsNights             4.0      6.0     116.00  
RoomNights                2.0      3.0     185.00  
SRHighFloor               0.0      0.0       1.00  
SRLowFloor                0.0      0.0       1.00  
SRAccessibleRoom          0.0      0.0       1.00  
SRMediumFloor             0.0      0.0       1.00  
SRBathtub                 0.0      0.0       1.00  
SRShower                  0.0      0.0       1.00  
SRCrib                    0.0      0.0       1.00  
SRKingSizeBed             0.0      1.0       1.00  
SRTwinBed                 0.0      0.0       1.00  
SRNearElevator            0.0      0.0       1.00  
SRAwayFromElevator        0.0      0.0       1.00  
SRNoAlcoholInMiniBar      0.0      0.0       1.00  
SRQuietRoom               0.0      0.0       1.00

In [8]:
# Summary statistics for non numerical variables
summary=ds.describe(include=[np.object],percentiles=None)   
summary=summary.transpose()  
summary.head(len(summary))

count  unique  \
Nationality          111733     199   
NameHash             111733  107584   
DocIDHash            110732  103480   
DistributionChannel  111733       4   
MarketSegment        111733       7   

                                                                   top   freq  
Nationality                                                        FRA  16516  
NameHash             0x15A713CE687991691A18F6CDC56ABE24979C73CF5D51...     75  
DocIDHash            0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   3032  
DistributionChannel                              Travel Agent/Operator  91019  
MarketSegment                                                    Other  63680

199 Nationalities, more than existing countries. We should be carefull for duplicates.  
We seem to have duplicates on Name Hashes and DocIDHashes , so we may have to do something about this.

In [10]:
# Check missing values
print(ds.isnull().sum())

ID                         0
Nationality                0
Age                     4172
DaysSinceCreation          0
NameHash                   0
DocIDHash               1001
AverageLeadTime            0
LodgingRevenue             0
OtherRevenue               0
BookingsCanceled           0
BookingsNoShowed           0
BookingsCheckedIn          0
PersonsNights              0
RoomNights                 0
DistributionChannel        0
MarketSegment              0
SRHighFloor                0
SRLowFloor                 0
SRAccessibleRoom           0
SRMediumFloor              0
SRBathtub                  0
SRShower                   0
SRCrib                     0
SRKingSizeBed              0
SRTwinBed                  0
SRNearElevator             0
SRAwayFromElevator         0
SRNoAlcoholInMiniBar       0
SRQuietRoom                0
dtype: int64


We must decide if we do imputation on Age or if we drop the columns. Maybe they are the no-shows?  
Missing document ID, must mean they did not present documentation at check-in wich is mandatory. So probably no shows? Decide on what to do.